### PR TITLE
Add missing device id for pytorch 2.8

### LINKED ISSUE #21095: UserWarning: No device id is provided
### Bug description

As of torch 2.8.0, [a warning is issued whenever no device id is provided via `init_process_group` or `barrier`](https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L4815).

This warning is issued at every step of training and clogs logs for long training runs.

This warning is issued due to [a call to `torch.distributed.barrier`](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/fabric/utilities/distributed.py#L228). This call doesn't pass `device_ids`, and the `ProcessGroup` passed [lacks a `bound_device_id`](https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L4804), which triggers the warning.

The warning is
```
~/minimal-no-device-id/.venv/lib/python3.13/site-packages/torch/distributed/distributed_c10d.py:4807: UserWarning: No device id is provided via `init_process_group` or `barrier `. Using the current device set by the user. 
```

Versions
```
lightning==2.5.3
torch==2.8.0
```

### What version are you seeing the problem on?

v2.5

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
import torch
from torch.utils.data import DataLoader, TensorDataset
import pytorch_lightning as pl
from torch import nn

# Minimal synthetic dataset: 100 samples, 10 features
X = torch.randn(100, 10)
y = torch.randint(0, 2, (100,))
dataset = TensorDataset(X, y)
dataloader = DataLoader(dataset, batch_size=16)

class DummyModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = nn.Linear(10, 2)
        self.loss = nn.CrossEntropyLoss()

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        logits = self(x)
        loss = self.loss(logits, y)
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)

if __name__ == "__main__":
    model = DummyModel()
    trainer = pl.Trainer(max_epochs=2, accelerator="auto")
    trainer.fit(model, dataloader)
```

### Error messages and logs

```
~/minimal-no-device-id/.venv/lib/python3.13/site-packages/torch/distributed/distributed_c10d.py:4807: UserWarning: No device id is provided via `init_process_group` or `barrier `. Using the current device set by the user. 
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

_No response_

cc @justusschock

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #21095

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by Borda — COMMENTED** [review]
can we have atest for it?

**Review by deependujha — APPROVED** [review]
Thanks for this PR. I've tested the changes on a **`T4 machine with 4 GPUs`**, and the code ran successfully. I can confirm that the original warning is no longer present.

File: src/lightning/fabric/strategies/ddp.py [hunk 1]
```diff
@@ -212,7 +213,14 @@ def _setup_distributed(self) -> None:
...
         self._process_group_backend = self._get_process_group_backend()
         assert self.cluster_environment is not None
-        _init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
+        kwargs: dict[str, Any] = {"timeout": self._timeout}
+        if _TORCH_GREATER_EQUAL_2_3:
+            kwargs["device_id"] = self.root_device if self.root_device.type != "cpu" else None
+        _init_dist_connection(
+            self.cluster_environment,
+            self._process_group_backend,
+            **kwargs,
```
Comment (Borda) [inline src/lightning/fabric/strategies/ddp.py]:
```suggestion
            **kwargs
```

File: src/lightning/fabric/strategies/fsdp.py [hunk 2]
```diff
@@ -663,7 +663,14 @@ def _setup_distributed(self) -> None:
...
         self._process_group_backend = self._get_process_group_backend()
         assert self.cluster_environment is not None
-        _init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
+        kwargs: dict[str, Any] = {"timeout": self._timeout}
+        if _TORCH_GREATER_EQUAL_2_3:
+            kwargs["device_id"] = self.root_device if self.root_device.type != "cpu" else None
+        _init_dist_connection(
+            self.cluster_environment,
+            self._process_group_backend,
+            **kwargs,
```
Comment (Borda) [inline src/lightning/fabric/strategies/fsdp.py]:
```suggestion
            **kwargs
```

File: src/lightning/fabric/strategies/model_parallel.py [hunk 3]
```diff
@@ -302,7 +302,14 @@ def _setup_distributed(self) -> None:
...
         self._process_group_backend = self._get_process_group_backend()
         assert self.cluster_environment is not None
-        _init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
+        kwargs: dict[str, Any] = {"timeout": self._timeout}
+        if _TORCH_GREATER_EQUAL_2_3:
+            kwargs["device_id"] = self.root_device if self.root_device.type != "cpu" else None
+        _init_dist_connection(
+            self.cluster_environment,
+            self._process_group_backend,
+            **kwargs,
```
Comment (Borda) [inline src/lightning/fabric/strategies/model_parallel.py]:
```suggestion
            **kwargs
```

File: src/lightning/pytorch/strategies/ddp.py [hunk 4]
```diff
@@ -200,7 +200,14 @@ def setup_distributed(self) -> None:
...
         self._process_group_backend = self._get_process_group_backend()
         assert self.cluster_environment is not None
-        _init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
+        kwargs: dict[str, Any] = {"timeout": self._timeout}
+        if _TORCH_GREATER_EQUAL_2_3:
+            kwargs["device_id"] = self.root_device if self.root_device.type != "cpu" else None
+        _init_dist_connection(
+            self.cluster_environment,
+            self._process_group_backend,
+            **kwargs,
```
Comment (Borda) [inline src/lightning/pytorch/strategies/ddp.py]:
```suggestion
            **kwargs
```

File: src/lightning/pytorch/strategies/fsdp.py [hunk 5]
```diff
@@ -260,7 +260,14 @@ def setup_environment(self) -> None:
...
         self._process_group_backend = self._get_process_group_backend()
         assert self.cluster_environment is not None
-        _init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
+        kwargs: dict[str, Any] = {"timeout": self._timeout}
+        if _TORCH_GREATER_EQUAL_2_3:
+            kwargs["device_id"] = self.root_device if self.root_device.type != "cpu" else None
+        _init_dist_connection(
+            self.cluster_environment,
+            self._process_group_backend,
+            **kwargs,
```
Comment (Borda) [inline src/lightning/pytorch/strategies/fsdp.py]:
```suggestion
            **kwargs
```

File: src/lightning/pytorch/strategies/model_parallel.py [hunk 6]
```diff
@@ -350,7 +350,14 @@ def _setup_distributed(self) -> None:
...
         self._process_group_backend = self._get_process_group_backend()
         assert self.cluster_environment is not None
-        _init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
+        kwargs: dict[str, Any] = {"timeout": self._timeout}
+        if _TORCH_GREATER_EQUAL_2_3:
+            kwargs["device_id"] = self.root_device if self.root_device.type != "cpu" else None
+        _init_dist_connection(
+            self.cluster_environment,
+            self._process_group_backend,
+            **kwargs,
```
Comment (Borda) [inline src/lightning/pytorch/strategies/model_parallel.py]:
```suggestion
            **kwargs
```
